study/shin5774/3-week/정적_static_개념.md

```diff
@@ -0,0 +1,59 @@
+# 정적(static)이란 무엇인가요?
+
+## 답변
+
+```
+정적이라는 의미를 가진 키워드로 이를 이용하여 정적 변수 및 메서드를 만들수 있습니다.
+정적 메서드와 정적변수는 프로그램 시작시 생성되어 프로그램 종료전까지 사용할수 있습니다.
+이 둘은 원래 method 영역에 저장되어 있기에 GC의 대상이 아니었지만
+자바 8버전 이후로는 Heap영역에 저장되기에 GC의 대상이 되었습니다.
```

Comment:
Method Area는 Heap 영역에 속하지만, ( PermGen )
예외적으로 GC의 대상에서 제외된다고 알고있습니다!

```text
The method area, 
also known as Permanent Generation (PermGen), is considered a part of Heap memory, logically, 
although the simpler implementations of JVM may choose not to garbage-collect it.

However, Java 8 removes the PermGen space 
and introduces a new native memory space named Metaspace.

Before Java 8, PermGen stores static members like static methods and static variables. 
Additionally, PermGen also stores interned strings.

As we’ve already discussed, PermGen space is replaced with Metaspace in Java 8, 
resulting in a change for memory allocation of the static members.

Since Java 8, 
Metaspace only stores the class metadata, and heap memory keeps the static members.
Furthermore, the heap memory also provides storage for interned strings.
```

정리하면
Java 8 이전 static 멤버들이 올라갔던 PermGen 은 논리적으로 Heap 메모리 상에 있던건 맞지만,
이를 구현으로 옮긴 JVM에서는 이 부분에 대해서는 GC를 적용하지 않았다고 합니다.

Java 8 이후부터는 PermGen이 MetaSpace로 옮겨졌지만,
static 멤버와 리터럴화된 문자열(String)은 여전히 Heap 에 존재하는 거 같네요.
또한, static 변수/메서드는 프로그램의 시작에서 생성되고 종료 시 할당이 끝나는 것이니
여전히 GC의 대상에서 제외되는 것 같습니다.

이 링크 한번 읽어보시면 좋을 것 같습니다.
https://www.baeldung.com/jvm-static-storage

Reply:
아 그렇군요!
static object가 heap으로 옮겨졌고 이것이 참조를 안하게 되는경우 GC가 발생한다고 이해했어서 (정확한 출처를 못찾겠네요 ㅠㅠ) 최초 작성한 대로 이해 한것 같네요!
자세히 찾아보니 GC의 대상은 아닌것으로 보여집니다.

좋은 첨삭 감사합니다!